### PR TITLE
Add support for very fast packet I/O using netmap(4) on FreeBSD

### PR DESCRIPTION
**Pull request by droe**
Initial support for sending and receiving packets using netmap(4) on FreeBSD.  Netmap is a framework for very fast packet I/O from userspace that uses pre-allocated ring buffers mapped into ZMap's virtual address space.  I have mainly tested on 1 GigE and 10 GigE links, both on netmap-aware and non-aware NICs, but perf advantage compared to the default BPF send code on even faster links than 10 GigE should be even more substantial.

Netmap has also been ported to Linux, support in ZMap is for FreeBSD only at this point.  I am not sure how interesting netmap is on Linux with the requirement to patch kernel sources, and how it compares in terms of performance to liburing etc.  Making the netmap send/recv code work on Linux too would mainly involve porting the link status and interface stats bits to Linux, so should not be a big deal in case there is interest.

There will be merge conflicts with #767.  In case you'll accept (a revised version of) this PR, you may want to wait with merging until after that has landed and I had a chance to rebase on top.

### CODE REVIEW

**Comment by zakird** [comment]
Thanks @droe, this looks awesome! Thank you as well for all of the other improvements over the past few weeks — I sincerely appreciate all the work you've put into ZMap recently. 

@phillip-stephens I think that we should look at what it would take to get this to work on Linux too, particularly given that we haven't yet seen superb performance from `io_uring`.

**Comment by phillip-stephens** [comment]
Yeah, thank you so much @droe for this and your other contributions! I'll give it a proper look over on Tuesday.

@zakird Yep sounds good, once we have that bare metal server with a 10 Gbit uplink, I can look into what performance `io_uring` is really giving and look into this if `io_uring` leaves anything to be desired.

**Comment by droe** [comment]
In case you want to test this on Linux:  I keep an untested branch with Linux support at `droe/netmap-linux` over in the `droe/zmap` fork.  Compiles on Ubuntu.  Have not run it on a netmap-enabled Linux kernel, so no idea how badly it is broken.  Keeping Linux support separate from this PR.

File: src/send-netmap.h [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+int
+send_run_init(sock_t sock)
+{
+	struct pollfd fds = {
+		.fd = sock.nm.tx_ring_fd,
+		.events = POLLOUT,
+	};
+	log_debug("send-netmap", "tx ring %"PRIu32": polling for POLLOUT", sock.nm.tx_ring_idx);
+	if (poll(&fds, 1, -1) == -1) {
+		log_error("send-netmap", "poll(POLLOUT) failed: %d %s", errno, strerror(errno));
```
Comment (phillip-stephens) [inline src/send-netmap.h:35]:
```suggestion
		log_error("send-netmap", "poll(POLLOUT) failed: %d: %s", errno, strerror(errno));
```

File: src/send-netmap.h [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+	struct netmap_ring *ring = NETMAP_TXRING(zconf.nm.nm_if, sock.nm.tx_ring_idx);
+	struct pollfd fds = {
+		.fd = sock.nm.tx_ring_fd,
+		.events = POLLOUT,
+	};
+
+	for (int i = 0; i < batch->len; i++) {
+		if (ring->head == ring->tail && poll(&fds, 1, -1) == -1) {
+			int oerrno = errno;
+			log_debug("send-netmap", "poll(POLLOUT) failed: %d %s", errno, strerror(errno));
```
Comment (phillip-stephens) [inline src/send-netmap.h:71]:
```suggestion
			log_debug("send-netmap", "poll(POLLOUT) failed: %d: %s", errno, strerror(errno));
```

File: src/zmap.c [hunk 3]
```diff
@@ -942,6 +956,98 @@ int main(int argc, char *argv[])
...
+	log_debug("zmap", "waiting max 10s for PHY reset to complete");
+	struct if_data ifd;
+	bzero(&ifd, sizeof(ifd));
+	struct ifreq ifr;
+	bzero(&ifr, sizeof(ifr));
+	strlcpy(ifr.ifr_name, zconf.iface, sizeof(ifr.ifr_name));
+	ifr.ifr_data = (caddr_t)&ifd;
+	for (size_t i = 0; i < 40; i++) {
+		if (ioctl(zconf.nm.nm_fd, SIOCGIFDATA, &ifr) == -1) {
+			log_fatal("zmap", "unable to retrieve if_data: %d %s",
```
Comment (phillip-stephens) [inline src/zmap.c:1037]:
```suggestion
			log_fatal("zmap", "unable to retrieve if_data: %d: %s",
```

File: src/send-netmap.h [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+		void *dst_buf = NETMAP_BUF(ring, ring->slot[ring->cur].buf_idx);
+		memcpy(dst_buf, src_buf, len);
+		ring->slot[ring->cur].len = len;
+		ring->head = ring->cur = nm_ring_next(ring, ring->cur);
+	}
+
+	if (ioctl(fds.fd, NIOCTXSYNC, NULL) == -1) {
+		int oerrno = errno;
+		log_debug("send-netmap", "ioctl(NIOCTXSYNC) failed: %d %s", errno, strerror(errno));
```
Comment (phillip-stephens) [inline src/send-netmap.h:88]:
```suggestion
		log_debug("send-netmap", "ioctl(NIOCTXSYNC) failed: %d: %s", errno, strerror(errno));
```

**Review by phillip-stephens — APPROVED** [review]
This looks really great overall, really appreciate the very digestible README.
Just a few very minor nits on formatting, I didn't see anything stand out in the core logic.

My FreeBSD VM doesn't have the `/dev/netmap` device by default, and based on [this](https://forums.freebsd.org/threads/freebsd-10-1-missing-dev-netmap.51562/), it looks like adding it would require compiling the kernel from source. IMO, if it's working on your BSD box, I'm good to merge if you are @zakird.

**Review by zakird — APPROVED** [review]
Looks good to me. I know that you had some nits about the code. Please merge when your'e good with state of thinsg @phillip-stephens !

**Comment by droe** [comment]
Re: netmap default availability, in case you're testing on an arm64 VM, netmap has only recently been added to the default config in [D43702](https://reviews.freebsd.org/D43702), and that change has not made it into a release yet; 14.0 still requires rebuilding the kernel (tl;dr: `git clone -b releng/14.0 --depth 1 https://git.freebsd.org/src.git /usr/src`, apply the diff in D43702, `make buildkernel && make installkernel`, reboot).  amd64 otoh had it in the default config for ten years now (11.0).

Will take care of the PR feedback later this week, thanks for reviewing.

**Pull request merged**